Move attribute and etag logic into Repository

bdemers opened the pull request:
Breaks up the current `update` method into new `update` and `patch` methods.
(Breaking change)

- ETag/Version logic must be managed by Repositories, except for the case of `ServiceProviderConfiguration` (which is likely stored in memory)
  - Implementations should set `scimResource.meta.version`, how this version is defined is up to the implementation
- The `UpdateRequest` class has been removed, the `getPatchOperations` method has been moved into a new class `PatchGenerator`
- Current examples have been updated and demo simple in memory patch application. e.g. Get/apply patch/persist.
  - Other implementations may want generate update queries directly from the list of PatchOperations (skipping intermediate get step)

bdemers committed:
Move attribute and etag logic into Repository

Breaks up the current `update` method into new `update` and `patch` methods.
(Breaking change)

- ETag/Version logic must be managed by Repositories, except for the case of `ServiceProviderConfiguration` (which is likely stored in memory)
  - Implementations should set `scimResource.meta.version`, how this version is defined is up to the implementation
- The `UpdateRequest` class has been removed, the `getPatchOperations` method has been moved into a new class `PatchGenerator`
- Current examples have been updated and demo simple in memory patch application. e.g. Get/apply patch/persist.
  - Other implementations may want generate update queries directly from the list of PatchOperations (skipping intermediate get step)

joshuapsteele commented on scim-core/src/main/java/org/apache/directory/scim/core/repository/Repository.java:
```diff
@@ -60,14 +63,31 @@ public interface Repository<T extends ScimResource> {
+  T update(String id, String version, T resource, Set<AttributeReference> includedAttributes, Set<AttributeReference> excludedAttributes) throws ResourceException;
+
+  /**
+   * Allows the SCIM server's REST implementation to update and existing
```
```suggestion
   * Allows the SCIM server's REST implementation to update an existing
```

bdemers replied:
Thanks!

joshuapsteele commented on scim-server/src/main/java/org/apache/directory/scim/server/rest/BaseResourceTypeResourceImpl.java:
```diff
@@ -382,35 +358,51 @@ private ScimException notFoundException(String id) {
-        .build());
-    }
-  }
+//  private void validatePreconditions(String id, EntityTag etag) {
+//    ResponseBuilder response = request.evaluatePreconditions(etag);
+//    if (response != null) {
+//      throw new WebApplicationException(response
+//        .entity(new ErrorResponse(Status.PRECONDITION_FAILED, "Failed to update record, backing record has changed - " + id))
+//        .build());
+//    }
+//  }
+//
+//  private EntityTag requireEtag(ScimResource resource) throws ScimException {
+//    try {
+//      return etagGenerator.generateEtag(resource);
+//    } catch (EtagGenerationException e) {
+//      throw new ScimException(Status.INTERNAL_SERVER_ERROR, "Failed to generate the etag", e);
+//    }
+//  }
+//
+//  private EntityTag etag(ScimResource resource) {
+//    try {
+//      return etagGenerator.generateEtag(resource);
+//    } catch (EtagGenerationException e) {
+//      log.warn("Failed to generate etag for resource", e);
+//      return null;
+//    }
+//  }
```
Still getting my footing in reviewing this PR. What's the reason for having these lines commented-out at the moment?

bdemers replied:
Good catch, i though I deleted all of the dead code, I'll take another pass through this PR!

bdemers replied:
Removed

joshuapsteele reviewed (commented):
Overall, this is looking good to me. Will definitely help us optimize database operations on our end when IdPs like Azure send us a bunch of single-user PATCH /Groups requests!

However, can we get some better comments/documentation about what, specifically, about the incoming PUT or PATCH requests determines whether the `update()` or `patch()` flows are used? In general, I tend to get lost in SCIMple when looking for the endpoint/controller-level logic. In this PR as well, I've missed (at least on a first pass) where the main logic is for SCIMple to decide whether to use `update()` or `patch()` based on the incoming request that it's handling.

bdemers commented:
> However, can we get some better comments/documentation about what, specifically, about the incoming PUT or PATCH requests determines whether the `update()` or `patch()` flows are used? In general, I tend to get lost in SCIMple when looking for the endpoint/controller-level logic. In this PR as well, I've missed (at least on a first pass) where the main logic is for SCIMple to decide whether to use `update()` or `patch()` based on the incoming request that it's handling.

Absolutely!

IMHO, the high level goal is that folks _shouldn't_ need to know any of the detail of the REST layer, and only need to implement the `Repository` API.

This will need to be expanded on in the javadoc and other docs, but here are some key points:
- There are two ways to modify an existing resource:
  - a replacement of a resource, `PUT` request, which calls `repository.update(...)` 
  - a partial update of a resource, `PATCH` request which calls `repository.patch(...)`
  
Maybe it would be helpful to have a table mapping the SCIM protocol requests back to the repository method:

(not completely accurate, mostly thinking out loud for how this might look)

| REST Endpoint | REST method | Repository method |
|----------------|---------------|--------------------|
| /Users | `GET`  | `repository.query()` |
| /Users/{id} ` | `PUT`  | `repository.update()` |
| /Users/{id} ` | `PATCH`  | `repository.patch()` |
...

As an aside, while talking about the javadocs, I'm not sure the Swagger/OpenAPI annotations in the REST layer add value to the user. (As the OpenAPI spec for SCIM could be defined by the body managing the RFC, or other project, that OpenAPI spec doesn't change unless the RFC is updated)
Anyway, I've wondered if removing it would make it easier to read/understand the code. (Or if I'm just over thinking it and that's not a problem at all)

joshuapsteele reviewed (approved):
Just a couple nits, really (typo, remove commented-out lines). Otherwise, this approach looks good to me. And it would really help us optimize group membership updates on our end!

erant10 commented on scim-server-examples/scim-server-jersey/src/main/java/org/apache/directory/scim/example/jersey/service/InMemoryUserService.java:
```diff
@@ -130,17 +134,21 @@ public ScimUser create(ScimUser resource) throws UnableToCreateResourceException
+
+  @Override
+  public ScimUser patch(String id, String version, List<PatchOperation> patchOperations, Set<AttributeReference> includedAttributeReferences, Set<AttributeReference> excludedAttributeReferences) throws ResourceException {
+    PatchHandler patchHandler = new PatchHandlerImpl(schemaRegistry);
```
would it make sense to inject PatchHandlerImpl to the Repositories instead of instantiating per request?

bdemers replied:
I just pushed an updated, and injected the PatchHandler, please take a look!

erant10 commented on scim-core/src/main/java/org/apache/directory/scim/core/repository/PatchGenerator.java:
```diff
@@ -0,0 +1,523 @@
+import java.util.*;
+import java.util.stream.Collectors;
+
+public class PatchGenerator {
```
I don't see this class used anywhere, so just making sure I understand - what exactly is the purpose of PatchGenerator? is it meant to be used optionally by anyone who wants to convert an `update` request into a `patch` request in their repo implementation? 

Perhaps a short comment on this class and its intended use could be helpful for future adopters?

bdemers replied:
💯 
Added Javadoc comments!

erant10 commented on scim-core/src/main/java/org/apache/directory/scim/core/repository/Repository.java:
```diff
@@ -60,14 +63,31 @@ public interface Repository<T extends ScimResource> {
-   * @param updateRequest The ScimResource to update and persist.
+   *
+   * @param id the identifier of the ScimResource to update and persist.
+   * @param version an optional version (usually used as an ETag) that be used to optimize update requests, may be compared against, the current {@code ScimResource.meta.version}.
```
I'm a little confused by the `version <-> etag` relationship. 
are they are equivalent?
if etag an http concept then it makes sense to distinguish the two, but just curious if there was more to it than that

erant10 replied:
also nit: that *can* be

bdemers committed:
Inject PatchHandler into exmaple constructors

Rename PatchHandlerImpl to DefaultPatchHandler (users of SCIMple could create different implementations)
Update examples to inject PatchHandler

bdemers committed:
Adding section in readme that maps HTTP requests to Repository methods.

NOTE: This section does not describe the supporting endpoints that SCIMple handles automatically

kjthorpe18 commented on scim-core/src/main/java/org/apache/directory/scim/core/repository/PatchGenerator.java:
```diff
@@ -0,0 +1,536 @@
+    }
+
+    /**
+     * There is a know issue with the diffing tool that the tool will attempt to move empty arrays. By
+     * nulling out the empty arrays during comparison, this will prevent that error from occurring. Because
+     * deleting requires the parent node
```
```suggestion
     * There is a known issue with the diffing tool that the tool will attempt to move empty arrays. By
     * nulling out the empty arrays during comparison, this will prevent that error from occurring because
     * deleting requires the parent node.
```

kjthorpe18 commented on scim-core/src/test/java/org/apache/directory/scim/core/repository/PatchGeneratorTest.java:
This covers ScimUser resources very well, but some basic tests for `ScimGroup` resources would be useful, too. 
- Modify `displayName`
- add member(s)
- remove member(s)
- replace member(s) - when should this be used over multiple add/remove operations? e.g. `1 replace` = `1 add` + `1 remove`

bdemers replied:
These are great!

The first 3 are pretty easy, the last one is a bit more complex, and will require a bit more work (and likely changes as to the JSON diff lib is invoked)

Since this logic didn't change in this PR (it was moved from a different class)
I'll create a separate issue to track this (and I'll stick the current tests i have in a new PR, as a starting point)

bdemers replied:
Created #427

kjthorpe18 commented on scim-core/src/main/java/org/apache/directory/scim/core/repository/PatchGenerator.java:
```diff
@@ -0,0 +1,536 @@
+        ++i;
+      }
+
+      if (patchOpType == PatchOperation.Type.REPLACE && (parseData.resourceObject == null ||
+        (parseData.resourceObject instanceof Collection && ((Collection<?>)parseData.resourceObject).isEmpty()))) {
+        patchOpType = PatchOperation.Type.REMOVE;
+        valueNode = null;
+      }
+
+      if (patchOpType == PatchOperation.Type.REPLACE && parseData.originalObject == null) {
+        patchOpType = PatchOperation.Type.ADD;
+      }
```
I see this is part what I mentioned in another comment - a replace operation actually being an add/remove depending on the original or target resource having `null`

kjthorpe18 reviewed (approved):
LGTM! Just a few small nits. Nice work here!

joshuapsteele reviewed (commented):
@bdemers Still think these changes look great and I'm eager to see them incorporated into the project! Who else do we need to bug so that they can review this PR? :)

bdemers commented:
@joshuapsteele All the reviews here are great!
I just cleaned up the git history a little, I'll merge it once CI passes!

joshuapsteele commented:
> @joshuapsteele All the reviews here are great! I just cleaned up the git history a little, I'll merge it once CI passes!

@bdemers Sounds good! @kjthorpe18 @zalbert02 and I are all eagerly anticipating this improvement! BTW, how can we get added as contributors to this SCIMple project? We're quite motivated to see it succeed and continue to be developed over time.

bdemers commented:
@joshuapsteele keep doing what you doing, I'll try to get that ball rolling!
In the short term, if you haven't seen this article take a look!
https://community.apache.org/contributors/
